Add migration guide for buildpack api 0.4->0.5 (#279)
buildpacks/docs: content/docs/reference/spec/migration/buildpack-api-0.4-0.5.md

+++
title="Buildpack API 0.4 -> 0.5"
+++

<!--more-->

This guide is most relevant to buildpack authors. 

See the [spec release](https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.5) for buildpack API 0.5 for the full list of changes and further details.

### Stack ID
The buildpack may now indicate compatibility with any stack by specifying the special value `*`.

### Character restrictions for process types

> For each process in `launch.toml`, `type` is now restricted to only contain numbers, letters, and the characters ., _, and -, so symlinks on both linux and windows filesystems can be created.

Reviewer: ```suggestion
For each process in `launch.toml`, `type` is now restricted to only contain numbers, letters, and the characters `.`, `_`, and `-`, so symlinks on both linux and windows filesystems can be created.
```

Author: Fixed as part of https://github.com/buildpacks/docs/pull/279/commits/030d158ab700a4ad8b91f90b8465f4b085408655.

### Cached layers
Cached layers are now part of the list of Layers Types for clarity.
 
### Override env vars
Override is now the default behavior for env files without a suffix.
It means that for the `/env/`, `/env.build/`, and `/env.launch/` directories, the default, suffix-less behavior will be `VAR.override` instead of `VAR.append`+`VAR.delim=:`.

### Decouple Builpack Plan and BOM
The Buildpack Plan and Bill-Of-Materials are now decoupled.

> The file in `/bin/build` that contains the Buildpack Plan entries for the buildpack is now read-only.

Reviewer: By "the file" we mean "the `<plan[ER]>` positional argument", right? I think we might need to rephrase this because it reads weird.

Author: @jkutner, thanks for your feedback! I rephrase this sentence as part of https://github.com/buildpacks/docs/pull/279/commits/030d158ab700a4ad8b91f90b8465f4b085408655.
I'm going to resolve this conversation but please feel free to reopen it or open a new conversation if you think the new phrasing is still unclear.

There are new `[[bom]]` sections in `<layers>/launch.toml` and `<layers>/build.toml` for runtime and build-time Bill-of-Materials entries respectively.
There is a new `[[entries]]` section in `<layers>/build.toml` for Buildpack Plan entries that should be passed to subsequent buildpacks that may provide the dependencies.

### exec.d

> The launch process now supports `exec.d` executables. Not like `profile.d` scripts that must be text files containing Bash 3+ scripts, the `exec.d` scripts do not depend on a shell but can still modify the environment of the app process.

Reviewer: ```suggestion
The launch process now supports `exec.d` executables. Unlike `profile.d` scripts, which must be text files containing Bash 3+ scripts, the `exec.d` scripts do not depend on a shell but can still modify the environment of the app process.
```

Author: Fixed as part of https://github.com/buildpacks/docs/pull/279/commits/030d158ab700a4ad8b91f90b8465f4b085408655.

+++
title="Buildpack API 0.4 -> 0.5"
+++

<!--more-->

This guide is most relevant to stack and buildpack authors. 

See the [spec release](https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.5) for buildpack API 0.5 for the full list of changes and further details.

## Stack author

### Stack ID

> Stack ID requirements have been removed from the buildpack specification.
> Each stack must indicate either a compatible stack and provide a valid stack ID, or compatibility with any stack by specifying the special value `*`.

Reviewer 2: ```suggestion
The buildpack may now indicate compatibility with any stack by specifying the special value `*`.
```

1. First sentence made me think that something is going away when in reality it's simply loosening a requirement.
2. It's really a buildpack that is essentially stating that it's _compatible_ with stacks. The only change is that we can now use `*` value to signify compatibility to _any_ stack.
3. Based on 2) this is still a Buildpack Author concern.

Author: I took the first sentence from the [PR description](https://github.com/buildpacks/spec/pull/155) but  I agree that your phrasing is much clearer. I'll go ahead and commit it.
Regarding 3), yes, you're totally right. I made a mistake at the beginning and forgot to fix it.

Author: Please see my changes in [this commit](https://github.com/buildpacks/docs/pull/279/commits/85fb1f1db86cb9637d7e6569af61afb42c91f00d).
I'll resolve this conversation. Please feel free to reopen it if you think that something is still missing.

 
## Buildpack author

### Character restrictions for process types

> `type` has been restricted to only contain numbers, letters, and the characters ., _, and -, and now symlinks on both linux and windows filesystems can be created.

Reviewer 2: It would be nice to specify the file in which this key exists.

Author: Please see my changes in [this commit](https://github.com/buildpacks/docs/pull/279/commits/85fb1f1db86cb9637d7e6569af61afb42c91f00d).
I'll resolve this conversation. Please feel free to reopen it if you think that something is still missing.

### Cached layers
Cached layers is a new layer type. A buildpack may specify that a `<layers>/<layer>/` directory is a cached layer by placing `cache = true` in `<layers>/<layer>.toml`.

> If a cache is provided, the lifecycle will store all cached layers after a successful build.

Reviewer 2: We had cached layers before, I'm not sure I understand what is changing.

Author: I'm sorry, it's a bad phrasing. The [spec](https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.5) indicates:
```
Adds "Cached Layers" to "Layer Types" to improve clarity
```
I'll reword this section.

Reviewer 3: This is a non-functional change, just making the spec clearer. So I think we could take it out of the guide.

Author: @natalieparellano, I agree with you so I've just pushed a commit that deleted it.
@jromero, @ekcasey, please let me know if you think I should add it back.
In the meantime, I'm going to mark this conversation as resolved.
